Create a proposal for Buffers API.

jbtk opened the pull request:
#### What type of PR is this?

/kind documentation

#### What this PR does / why we need it:
Adds proposal for Buffers API

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

jbtk commented:
The initial discussion about this proposal was on a doc: http://docs/document/d/1bcct-luMPP51YAeUhVuFV7MIXud5wqHsVBDah9WuKeo?tab=t.0

(there are still unresolved conversations there)

jbtk commented:
Update: 
1. added k8s resource quota handling to proposal
2. after some discussions I am keeping the proposal to have a separate controller. When implementing we will reconsider embedding pod spec (due to embedding and validation that needs to be copied and maintained)

towca commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+}
+
+type BufferStatus struct {
+    PodBasedBufferSpec   PodBufferCapacity
```
Maybe this should be a list for future-proofing? I imagine we might have buffers that would want to translate into non-homogeneous Pods in the future.

towca commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+
+### What is the responsibility split between the Buffers controller and autoscaler?
+
+The controller would own the logic of translating the buffer configuration into
```
Could you rewrite this part more concretely, in terms of the new proposed objects? E.g. that autoscalers would only react to `BufferStatus.PodBasedBufferSpec` etc.

More generally, it's not clear from the proposal how the new objects are supposed to be interacted with by either a Node autoscaler or the Buffer controller. Maybe instead of having this section we should have separate sections explaining how both sides should use the API? Including which fields they consume and set, what kinds of objects the reference Buffer controller can translate, etc..

towca commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+
+### How to directly define cluster capacity?
+
+Currently the only option to define the cluster capacity is to define a PodSpec.
```
Shouldn't this section be updated now that the proposal is sent for review? Also I'm not sure I understand the first sentence, shouldn't it be something like "the only option to consume Node capacity in the cluster is via a Pod"?

towca replied:
Thanks! I still think that the use of "defining capacity" here is misleading. Pods/reservations don't define capacity, they consume capacity defined by Nodes/ResourceSlices.

towca commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+    but jobs that start and finish. It is impossible to start them in advance to
+    keep the space in the cluster.
+
+### Why does NodeClassBufferSpec not have relative size?
```
Outdated names used in this section as well

jbtk replied:
Done

towca commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+## Buffers dedicated for workload
+
+We are considering adding buffers dedicated for a workload in next iterations.
+For full context this is how they would differ from the v1 milestone:
```
The initial API version will be `v1alpha1`, right?

jbtk replied:
Since the numbers here do not matter I changed the test to say "first iteration"

towca commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+    // Depending on the type of referenced object one of the configurations below will be used:
+    //   - for a crd implementing scale subresource (like deployment, job, replicaset)
+    //     the ReplicaBasedBufferSpec will be used
+    //   - for GKE CCC or Karpenter NodePool the NodeClassBufferSpec will be used
```
I'd avoid cloud-provider-specific terms in the object definitions/comments, maybe we can use "cloud-provider-specific Node class" instead?

It's also not clear to me how CCC or NodePool are different from a "custom CRD" here, i.e. what proposed logic will ensure that "NodeClassBufferSpec will be used"? Will the reference controller implementation be able to translate either of them?

jbtk replied:
Changed as we moved target ref to be inside the capacity object where applicable.

towca reviewed (commented):
The overall idea and direction looks good to me, but IMO we should clarify some fine details in the proposal.

@jackfrancis @gjtempleton Could you give the proposal a high-level review?

@jonathan-innis Could you review the proposal? It's translated from the doc we've already discussed.

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+
+# Proposal
+
+A `Buffer` CRD (autoscaling.x-k8s.io) is added to represent spare capacity
```
+1 for `NodeBuffer`, though I'd also throw `NodeHeadroom` into the mix, which I like a bit more :).

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+If the user uses autoscaling the cluster size will be adjusted to the number of
+currently running pods. This means that often when a new pod is created its
+scheduling time will need to include the time of creating a new node and
+attaching it to the cluster. Many users care deeply about pod scheduling time
```
There are a nontrivial amount of writes (events, etc) associated with evicting, deleting, creating, and scheduling pods. ETCD is pretty sensitive to writes -- I suspect there are pretty huge perf benefits to this if you were to measure it.

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+    additional pods that are representing a buffer (for example treating them as
+    virtual pods) so that the capacity in the cluster becomes available.
+
+### How do we determine the pod spec for targets using scale subresource?
```
Why not use a pod template spec inline? Imagine the case where you want to leave room for a CI job, but the job doesn't exist in the cluster yet, so you have nothing to reference.

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+    require more implementation work within the cluster autoscaler in order to
+    support virtual pods with DRA resources.
+*   More features that further help users to manage the buffers:
+    *   Buffer schedule - option to define a buffer that is active only during
```
If Buffer supported the scale subresource, you could hook it up to things like https://keda.sh/docs/2.11/scalers/cron/ to achieve this use case.

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+        schedule quickly on the cluster. Tests are mostly run during business
+        hours of my organization and I do not want to pay for the extra capacity
+        at night or over the weekend.
+*   Buffer dedicated for workload: mark the buffer as dedicated for pods defined
```
@jonathan-innis and I discussed a concept of `fulfilledBy: metav1.Selector` that does exactly this. I think this conflicts with `RecreationStrategy`. They solve the same problem, right?

jbtk replied:
For now I am removing the whole recreation strategy and fulfilledBy. I will come back with this problem in the next PR to this AEP. For now I want to get this to a state where we can start implementing what we already know. I will come back with a pr only for this in a couple of days.

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+  name: testing-capacity
+  namespace: my-namespace
+spec:
+  targetRef:
+    apiVersion: apps/v1
+    kind: CustomComputeClass
+    name: my-custom-class
```
Why not use podSpec to define the buffer here?

liggitt replied:
> since writing this doc I have heard that embedding full pod spec is quite of a headache for the future

+100 ... leaning on references to the actual PodTemplate API instead of inlining will get you a ton of built-in validation behavior and support for all future Pod fields automatically and for free

jbtk replied:
@liggitt - actually there are two places where we need some pod spec - one in the buffer definition - here after learning about all the issues I am leaning towards pod template reference. However we also have a pod spec in the status field (we are translating the buffer config into a pod spec that will be used by autoscaler to know what to provision). Here I think we have three options:
1 pod template - I am not sure about this one because of the need to create something every time when buffer changes and also because of the user access
2 embedding full pod spec
3 creating our own subset of pod spec fields that are used by scheduler/autoscaler to place a pod and using that instead of the full pod spec

For 2 we would simply have full pod spec (that is still embedded in the CRD definition) - at least it would be possible to just script copy pasting of the fields etc. 
For 3 we have a manually curated set (so a bit more maintenance to keep it up to date), but significantly less fields. 

If you have any thoughts which one is better I am happy to discuss. As I wrote above - when writing this design I was not even aware how problematic is embedding a pod spec in general so happy to learn about best practices.

jbtk replied:
Synced with @liggitt - it seems that the best solution for buffer status will be to use pod template there and to save generation number to ensure that user changes will not influence what buffers we provision.

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+
+    // Initially only active stand by meaning scaling up the cluster will be available.
+    // In the future the cloud providers will be able to offer their own buffer types.
+    Type string
```
If there's only one option, do we need this field at all? It's not clear to me what other types might be.

jbtk replied:
Changed to ProvisioningStrategy and added comment about default behaviour when unspecified.

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+    // Depending on the type of referenced object one of the configurations below will be used:
+    //   - for a crd implementing scale subresource (like deployment, job, replicaset)
+    //     the ReplicaBasedBufferSpec will be used
+    //   - for GKE CCC or Karpenter NodePool the NodeClassBufferSpec will be used
```
For node-centric headroom, I suspect that this will be implemented in the NodePool API in Karpenter, and not part of this buffer API.

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+    // How the buffer should be maintained over time:
+    // - KeepEmpty - upscale the cluster to keep the empty buffer on top of the running workloads
+    // - KeepMinimum - upscale the cluster and let workloads fill in the buffer as needed
+    RecreationStrategy string
```
KeepMinimum only makes sense in the context of a selector (i.e. pods that selectively fill the buffer). I think this Enum should probably be replaced with something like `fulfilledBy: metav1.Selector`

jbtk replied:
One shot was in the original proposal and was removed exactly for the complexity reason that you describe. The way to solve this would be that one shot is kept all the time, but its size changes as the incoming pods matching the buffer are created (in case of no fulfilled by it means that any pods that can be on the same nodes). This would require an admission hook to adjust the buffer size as the size goes down.

However the KeepMinimum is recreated but it accounts for the pods that are already in the cluster and match the same nodes as filling it in. I imagine the logic to look like this:
1. iterate over all the nodes in the cluster
2. check whether the buffer could be on this node
3. if yes count the running pods as already a buffer space (they are "filling in" the buffer)
4. adjust the buffer status to reflect how many empty chunks to keep in the cluster

Buffer controller would need to do it periodically and update the number of pod chunks that should be kept in the cluster.

If we ever add fulfilledBy in 3 we would only count pods that are matching the fulfilledBy.

jbtk replied:
Also some context:
originally I was thinking of keeping a boolean whether the buffer should be recreated or not and separate field of fulfilledBy, but later I realized that this will be not enough as if we want to add in the future a buffer that would be implemented as a cluster reservation (once/if sig-scheduling decides to implement scheduler level reservation concept) we would have a problem to express information like:
dedicated for a workload (only this workload can use this reserved space) but kept on top of the running workloads making 

Now that I have a second look at this and after removing one shot we could probably simplify, but I am afraid that it would be temporary and it feels like having a separate field for information about filling the buffer from how the buffer recreates would be more future proof (we can more easily extend it in the future with more enum values). Happy to get more feedback though

towca replied:
I agree with @ellistarn here, IMO `KeepEmpty` as the default behavior and optional `fulfilledBy` covers all relevant use-cases apart from one-shot and is much easier to understand than `KeepMinimum`. If we are to support one-shot in the future, we can always add a strategy field for it when we get to that point.

I'd drop `recreationStrategy` and start with just the `KeepEmpty` behavior, or with optional `fulfilledBy` instead of `recreationStrategy` if it's important to get that in alpha.

jbtk replied:
As commented elsewhere - removing from this PR to get approvals and start implementation work. Will come back after the long weekend with PR updating this AEP only for this where we will be able to discuss in detail :)

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+    // One of the following must be set
+    // +optional
+    Replicas *ReplicaBufferCapacity
+    // +optional
```
If you have a pod-centric headroom concept, why do you also need a replica based and nodeclass based concept?

ellistarn replied:
Is it right to think of `NodeClassBufferCapacity` as syntactic sugar? Everything you can express with it you could also express with `PodBufferCapacity`, it's just less/simpler config?

As a side-note "NodeClass" might be as bit confusing for Karpenter users, as it's an orthogonal concept to Karpenter's NodeClass concept (maps better to its "NodePool" concept). 

WDYT about calling this `Nodes` for parity with what you have w/ `Pods`. Each is node-centric and pod-centric buffering, respectively.

towca replied:
> Maybe just Resources and the type called ResourceBufferCapacity? @towca - wdyt?

LGTM!

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+
+##### Does this feature depend on any specific services running in the cluster?
+
+Yes, autoscaler and buffer controller (the latter run as part of cluster
```
Done

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+    // pods to create buffer not exceeding the total size above
+    PerChunkResources *ResourceLimits
+    // Optional if more detailed limitations than target object ref on its own is not enough.
+    NodeSelector map[string]string
```
You might want to expand this to https://pkg.go.dev/k8s.io/api/core/v1#NodeSelector, which is a more flexible/expressive version of map[string]string. e.g. instance-type in [a, b, c]

ellistarn replied:
It's used in a few of places, but most notably, it's used in NodeAffinity.

* https://pkg.go.dev/k8s.io/api/core/v1#NodeAffinity
* https://pkg.go.dev/k8s.io/api/core/v1#VolumeNodeAffinity

Semantically, you can think of it as a superset of the `map[string]string` variant. e.g.

```
- key: node.kubernetes.io/instance-type
- operator: In
- values: ["m5.large"]
```

is equivalent to 

```
node.kubernetes.io/instance-type: "m5.large"
```

But you can do more powerful things like 

```
- key: node.kubernetes.io/instance-type
- operator: In
- values: ["m5.large", "c5.large"]
```

which gives an OR semantic, "either m5.large or c5.large is fine".

---

From a buffers perspective, it could be useful to say "the capacity can be in us-west-2a or us-west-2b, but not us-west-2c".

adrianmoisey commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,583 @@
+    The reference implementation would be provided within the cluster autoscaler
+    repository, but it is expected that most integrations will depend on it as a
+    library rather than full implementation in order to support
+    autoscaler/provider specific configurations like CCC (in GKE) or NodePool
```
Is CCC "Custom Compute Classes" ? Can it be mentioned in full for those not too familiar with it?

jbtk replied:
Done

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,583 @@
+    //   - for autoscaler specifi node class the NodeClassBufferCapacity will be used
+    //   - in case of no target object the user will be able to directly define needed capacity via pod definition
+    //   - … - custom crd if the user runs their own buffer controller that can handle this target
+    TargetRef *TypedLocalObjectReference
```
Because the behavior depends on the relevant capacity structs below, it might be more straightforward to reason about if this reference was pushed into those structs, e.g.
```
spec: 
  capacity:
    object:
      replicas: ...
      ref: ...
```

>     //   - in case of no target object the user will be able to directly define needed capacity via pod definition

In this case, you'd just have:
```
spec: 
  capacity:
    pods: ...
```

towca replied:
+1, IMO this makes for a much clearer API

jbtk replied:
Done, @towca - please have a look as I am not sure whether it is ok that ReplicaBufferSpec has now one required field and two optional that have one of semantic. If this is not ok - any ideas for the name of the subtype for swith field? :P

jbtk replied:
Consulted with @liggitt - we already have quite a few examples in the k8s objects that do that so it should be ok (easy to search with string "// +oneOf=")

towca commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,567 @@
+*   User can define spare capacity in the cluster that should be available for
+    the user workloads
+*   User pods can use the available capacity without any configuration changes
+    (strictly additive, can be managed by the user who owns the pod as well as
```
That makes sense, I wonder if it'd be slightly clearer without the "strictly additive" part.

towca commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,583 @@
+
+#### Buffer controller
+
+Base on the full Buffer spec writes the status including conditions (to mark it
```
Shouldn't we define the conditions in this proposal?

jbtk commented:
@ellistarn as FYI

After an offline review with @towca 

- Changed name Buffer to NodeAutoscalingBuffer
- Changed the Type field to ProvisioningStrategy and added information that if not specified the default behavior will be active capacity (meaning provision more space in the cluster). The field can be still used for provider specific buffer types if needed
- Renamed NodeClassBufferCapacity to ResourceBufferCapacity and all related fields
- Moved targetRef into ReplicaBufferCapcity and ResourceBufferCapacity
- Added documentation for conditions that will be a starting point for implementation. More conditions can be added in the future if needed.

I also, for now, removed the recreation strategy. There is still one important usecase* that will require some configuration but in order to unblock implementation we decided to remove it for now and reintroduce in next pr to the same AEP and in the meantime start implementation of what we already have. All of these ideas assume that the implementation of logic how many of the capacity to provision will happen in the buffer controller - we need to get the config options right though for the API to be simple and expressive.

We still did not decide whether it is better to embed full pod spec or only a subset of fields related to scheduling. TBD.


* Usecase being: I am a user who has jobs running. I know how much they usually take (or I have a reservation of a certain size that I pay for anyway) and want to keep buffers up to this capacity.

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,583 @@
+    name: my-custom-class
+  type: buffer.x-k8s.io/active-capacity
+  recreationStrategy: keep-empty
+  nodeClassBufferSpec:
```
updated

ellistarn commented:
Thanks @jbtk, I'll review again soon when I get the chance.

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,578 @@
+  replicaBasedBufferSpec:
+    targetRef:
+      apiVersion: apps/v1
+      kind: Deployment
```
I wonder -- does it make sense to support daemonsets here, too? There are cases where daemonsets can't safely upgrade when there is no room on the node, so this could force some space on each node.

ellistarn commented:
Thanks @jbtk,

For 1, I get your point about the potential confusion. Any thoughts to my suggested name: NodeReservation? This one especially makes sense to me if we decide to support node hibernation with this API. It might also be worth exploring a "Pod" prefix (e.g. PodBuffer, PodReservation) for these names to help clarify the subject. 

For 2, apologies -- this thought was half baked and the last sentence incomplete. My thought was that arbitrary targets might require customized RBAC, but the design already assumes that each provider will have a separate implementation w/ separate RBAC, so I think this is solved.

---

For 3/4, I think these can be summarized as "speculative features that try to address future extensibility". This is philosophical, so I can understand if you feel differently, but I'd recommend that we delay the addition of these additional fields/features until the use case is concrete. 

For example, if CAS/GKE/Karpenter need targetRef to support something that cannot be easily supported with NodeSelector, let's add a follow-on that builds the extensibility and maps it directly to the use case.

In my experience, building speculative extensibility carries risk that the feature will not fully meet the future requirements once they're known. This can lead to awkward/unnecessary fields that are confusing to customers. We can easily add to APIs with new features, but removing from APIs is much more costly if not impossible. It's valuable to think of potential extension points (e.g. our discussion on fulfilledBy), even if we don't implement them immediately.

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,578 @@
+  name: my-deployment-buffer
+  namespace: my-namespace
+spec:
+  replicaBasedBufferSpec:
```
This was an original proposal, but I got feedback about two value options not being the best and additionally when one field works with one option and not the other. This embedding was proposed during the review here: https://docs.google.com/document/d/1bcct-luMPP51YAeUhVuFV7MIXud5wqHsVBDah9WuKeo/edit?disco=AAABjQp5l44

towca replied:
I don't personally have a strong preference here, IMO we should follow Kubernetes API best practices here. Do we know of any other similar fields in the K8s API?

@jackfrancis @gjtempleton WDYT?

jbtk commented:
@ellistarn 


1. reservations are considered to be added on the scheduler level (some docs: [older one](https://bit.ly/k8s-reservations), newer approach that was rejected, but may come back in a similar form: https://github.com/kubernetes/enhancements/pull/5149). Ultimately we do not reserve the space - only create it and keep it up (it has no guarantees that it will not move between nodes etc as scheduler is unaware of this space being taken). As I said - I do not want to create an impression that Buffers operate on a node level as this is not the intention of this api. I would rather go with ResourceBuffer, but "resource" is again a pretty ambiguous word. Though the resource word matches the resources fields in the pod spec which makes it potentially a match.

3. For target ref in resources capacity - this is not something I have a strong feeling about - I think that for any custom object that is a single thing (GKE custom compute class or any other custom CRD that users would like to write custom buffers for) it is much more convenient to directly point to that object rather than reference it by node selectors (for CCC it happens to be possible, but for custom CRD not necessarily making it almost the same as pod based config with only a feature of splitting the resources between chunks of the buffers). Also with direct ref we could automatically use it for owner ref metadata field. Since this is not required as CCC can be referenced by a node selector (not direct and nice, but working) I can remove this if others also feel that this is not a good idea. 

4. As for the type - this is something that has very concrete ideas (like stopped vms or resizeability) and does not have a workaround other than an annotation that has a risk of buffer controller not supporting these and acting as if buffer was for active capacity. This one I would rather leave as is. 

Since 3 and 4 is a matter of opinion it would be great for others to chime in. @towca - are you the right person or is there someone else who may have an opinion here?

towca commented:
@jbtk thanks for addressing my comments! I feel like we're very close to alignment, let's finalize this so that we can start implementing something.

IIUC these are the contentious parts that are preventing us from approving the proposal:

1. The `Buffer` name.
    * We need to distinguish that it's a Node-autoscaling-related (vs workload-autoscaling) object, so `Buffer` is too ambiguous.
    * `NodeBuffer` suggests that the Buffer is "Node-centric" (i.e. "Buffer _consisting of_ Nodes"), while the semantics are actually more "workload-centric" (i.e. "Buffer _for a_ workload"). I agree that this is confusing and would prefer to avoid this option. `NodeHeadroom` has the same problem.
    * Including `Reservation` in the name is IMO confusing because nothing will actually be "reserved" for Buffers until we implement the `fulfilledBy` semantics. Also there's the name collision with scheduler reservations that are being discussed separately, as @jbtk mentioned.
    * So we need a name that is accurate, distinguishes from workload-autoscaling, doesn't suggest that it's "Node-centric", and doesn't have collisions with other k8s concepts. IMO `NodeAutoscalingBuffer`, or `ResourceBuffer` both satisfy all points (the accuracy part is of course subjective).
    * IMO `CapacityBuffer` would be slightly better than `ResourceBuffer`. I'd also like to resurface my early suggestion of `node-autoscaling.x-k8s.io/Buffer` which also seems to satisfy all requirements.

1. Whether to include `TargetRef` in `ResourceBufferCapacity`.
    * @jbtk Do I understand correctly that `ResourceBufferCapacity` is supposed to mean "buffer for total amount of resources, and the arbitrary CRD just specifies where to create the resources"? If so, I'd name the field something else than `TargetRef` since the "target" here seems to be the total resources+chunk config (following the `ReplicaBufferCapacity` pattern where "target" is a workload). Maybe `NodeSelectorRef`/`NodeSelectorObject` or something similar?
    * @ellistarn Having the reference to the object gives us better observability (we could use it in `kubectl` etc). It's also easier for the controller to identify the object to translate (imagine if e.g. we have different default chunking strategies depending on the object). NodeSelector can work for this, but it's imprecise and essentially hacking the semantics.
    * Maybe it would be clearer if we put NodeSelector and the Ref in a separate subobject meant to select where to create the capacity? @ellistarn could you expand on what is your concern here?

1. The format of replica count/percentage in `ReplicaBufferCapacity`.
    * As mentioned in the other comment, for this one I'd defer to prior art in K8s API, or an opinion of a K8s API expert.
    * Maybe we can get Tim to do another review pass?

1. Whether to include the `ProvisioningStrategy` field.
    * As @jbtk mentioned, we do have concrete use-cases in mind that we can talk about right now. The alternative is coming back to this discussion in a couple of months. I'd prefer to avoid that, as any iteration seems to have a non-trivial overhead of kicking off and maintaining the discussion.
    * We can focus on the simple use-case that @jbtk mentioned - "stand-by"/"stopped" VMs that can be quickly resumed. Having the `ProvisioningStrategy` field allows the user to specify that the Buffer should be created using such VMs.
    * @ellistarn I agree that we might need more discussion and design later when the use-cases are more concrete, but it seems safe to assume that we'll need to at least be able to distinguish from the default behavior. Which is all that the field does. Are you concerned that we'd want a completely separate API for this use-case?

A lot of this is subjective so it'd be good to hear more opinions. @gjtempleton @jackfrancis @x13n could you weigh in on the points listed above?

jbtk commented:
@towca @ellistarn 

Yesterday I met with @liggitt to discuss embedding the pod spec. Following his advice I decided to get rid of embedding and the the buffer status we will use the pod template and save the generation number to ensure that the template was not modified by the user. This way we will be able to ensure that we do not over provision if we reach resource quota limits. The pod template will have the owner reference set so when buffer is deleted it will be deleted as well.

I also used the opportunity to consult on the IntOrString type. @liggitt told me that in newer APIs we try to avoid these multipurpose fields as they make the processing harder (harder to do validation, some other fields work with one value but not the other). Taking this into account I would rather leave these fields as they were.

jonathan-innis commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+    TargetRef *TypedLocalObjectReference
+    // One of the following must be set:
+    // +optional
+    Exactly *int
```
```suggestion
    Replicas *int
```

I get that the top-level thing is called replicas and that's what's causing us to not be able to call this field "replicas", but that (to me) suggests that we've named the struct the wrong thing. Replicas is used so heavily in K8s to reference an integer field that it feels odd to have it be a stanza rather than an integer in this API

jbtk replied:
Changed.

jonathan-innis commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+    MaxCount *int
+}
+
+type ResourceBufferCapacity struct {
```
With that API above in mind, I'm worried that there is scope creep that is happening here and we aren't addressing a small set of user stories first and then expanding out as we get feedback on this thing. For instance, maybe it makes sense just to create another kind called `ResourceBuffer` that's strictly different from `PodBuffer` since the concepts might not intermingle well

The other thing to consider here is that the "resource-based buffer" is just an ease-of-use thing. Users can totally orchestrate a resource-based buffer by just crafting a pod themselves -- granted, that is painful, but there is a workable workaround with this API for doing something like that that we could build on

jackfrancis replied:
The problem with this is that "4 extra nodes" doesn't mean anything. 4 large nodes? 4 small nodes? 4 nodes, each of which has a different cpu|mem|storage profile?

I do understand the use case, but I'm not sure there's a convenient slot to solve for it it using the core Kubernetes APIs, as there's no such definition of "node replica set". I wonder if "node buffer" needs to be something that is solved inside each particular node autoscaler's solution domain itself (because there is no standard Kubernetes API definition of a node replica set, different autoscalers will have different definitions of what that means).

To solve for the "a group of workloads running on a single group of nodes" use case we could add a buffer on a pod label match, which those group of workloads could agree upon.

jbtk replied:
Our thinking was that the resource buffer capacity would cover only basic usecases for generic classes of nodes.

Use case: I am a cluster admin, I am offering my cluster users 3 compute classes to deploy various workloads (that I do not have much knowledge about). I will provision extra x cores withing the highest priority class so that the workloads deploy faster.

Asking admin to define the chunks up front creates friction - they do not know what workloads will be running. In the initial proposal we will use some constant, but  in the future we could (this is speculative) make the chunking logic take into account what is currently running in the cluster. With pod spec we do not have any room for improvement - we just force the admin here to provide some values.

As discussed in DM on Slack with Jonathan - we are discussing internally with our product team whether this is actually needed (in the first version or in general).

jbtk replied:
As for scheduling constraints - I would imagine that we agree on a very basic set of things that are possible through this configuration (currently I am proposing only resources and node selectors). In more complex cases this would need to fallback to the pod spec in the pod template reference.

jonathan-innis commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+type NodeAutoscalingBufferStatus struct {
+    PodBufferCapacity   PodBufferCapacity
+	PodTemplateGeneration int	
+    Conditions           []BufferCondition
```
```suggestion
    Conditions           []metav1.Condition
```

Consider using [`metav1.Condition`](https://github.com/kubernetes/apimachinery/blob/v0.33.2/pkg/apis/meta/v1/types.go#L1604) since this is already a well-known concept

jbtk replied:
Confirmed in the sig-architecture slack that metav1 is a better way to go. Fixed.

jonathan-innis commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+This design has two main advantages compared to embedding all the logic in the
+library that would be used in autoscaler code:
+
+*   Responsibility split: it splits the responsibility of understanding buffer
```
For the second bit here, I'm not quite sure what "picked up" means -- even if the controller has reconciled the new pod definition into the status, that still doesn't mean that the autoscaler has seen the new pod. This problem exists whether you use API to store the state or store the state internally in memory (the library way that you are talking about above) so I don't really see this as a tangible benefit of one over the other

jonathan-innis commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+    the cluster based on the buffer spec that was provided. Also if for example
+    the target deployment changes it will be visible whether the change was
+    already picked up.
+*   Extendability: this design will allow for the users to either opt in the
```
I'm worried that we are trying to do this but aren't really sure how this would be handled concretely -- there isn't a strong use-case that I can think of for having divergent behavior on the pods that you gen from a given template -- I actually think that this makes things more confusing for a user who has to reason about what kind of output that they will get based on the components that they install

jonathan-innis commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+
+### Why not?
+
+*   Encapsulate the translation logic in a single controller that can be
```
Anytime I see the phrase "if the need arises", that means there isn't a need right now and therefore we are overshooting the mark with what we are doing for today. Since it's always possible to add this later and it's not possible to go back once it's added, I'm still strongly in favor of us starting with the library-based approach before we go with the status-based approach

jbtk replied:
If the need arises meaning: if the user wants to implement their custom logic for example of having buffer for their own custom CRD.

jonathan-innis commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+    with a user extended controller as long as it gets the list of buffer pod
+    specs from the buffer objects. This allows users to define buffers for their
+    custom CRDs without changing the autoscaler in any way
+*   Debuggability - splitting translation from provisioning of the capacity
```
This is the strongest argument to me for putting the result of the pods that we are going to provision for in the status -- now, I still think it won't be the easiest thing to read in the world because we're going to be putting a full pod spec into the status of something, but this is the most convincing thing to me

jbtk replied:
In the end we will not be putting the spec to avoid pod spec embedding - the buffer will create a pod template that it will reference.

jonathan-innis commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+Additionally adding a feature to buffers (like scheduled buffers or a new buffer
+type) would require again touching every single object.
+
+## No controller
```
I think I've mentioned this in a few previous discussions that we have had, but I _really really_ think that we need some concrete use-cases/reasons that we think the controller-based solution is significantly better than the no-controller solution. I can't think of strong extensions that would be useful to have the pod-based approach in this moment so I'd love to hear some if you are thinking of any

jbtk replied:
I think that there are two strongest reasons for me:
- split of the responsibility including resource quotas that are today totally outside of the autoscaling scope (as they are fully handled on admission)
- once reservations are delivered in scheduler we will be able to offer buffers that are backed by scheduler reservations without even autoscaler involved. 

Without controller we are limiting people to only buffers that are pulled in as a part of their autoscaling solution and I do not see a big gain from getting rid of the controller. Can you say what concerns you the most about the separate controller (that can run as a subprocess of the autoscaler/karpenter not necessarily as a separate pod in the cluster)?

jonathan-innis commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+    appropriate reservation underneath. Binding this logic to autoscaler does
+    not feel right.
+
+Note that while some of these arguments are not very strong the cost of having a
```
There's a lot of arguments here around "the cost of doing this is low" -- agreed that the thing is easy to build probably, but it's not just about this -- there's a cognitive and maintenance cost to adding these things in and I'd rather take decisions that we can iterate on than decisions that are more one-way

jbtk replied:
Why do you think that the cognitive cost is higher to the users between a controller that is a subprocess in autoscaler vs a library that is called within autoscaler? I see one thing - the fact that buffer is in multiple states. But even in the library the buffer can fail in multiple ways so the user will need to understand that first the buffer is translated and only later it is provisioned.

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+    PodBufferCapacity   PodBufferCapacity
+	PodTemplateGeneration int	
+    Conditions           []BufferCondition
+    LastFullyProvisioned *metav1.Time
```
We will not have a condition for a fully provisioned buffer. We only have a condition for the fact that the cluster is a scaling for this buffer.

Why?
We could have an additional condition for fully provisioned, but this information is very volatile - it is enough for a single pod to get scheduled in this space for it to be stale (and we do not have a good way to set this condition to false right away). Therefore I did not want to create a condition that looks like a way to check whether a buffer is fully provisioned but in fact only the timestamp there is useful and decided to go with just the timestamp field.

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+        apiVersion: apps/v1
+        kind: Deployment
+        name: my-deployment
+      percent:
```
The tricky part with such a setup is that min and max apply only if the percent is used and not when a constant number of replicas is set - splitting it into separate object clearly indicates that (though I agree that double percent is weird). The min value is mostly for cases where users say that they want to have 10% of their workload, but at least 1 replica. This is dynamic as this is a percent of the current size of the workload (for example deployment).

We cannot put replicas config directly under capacity as this is an object with three options: replicas, resources and pods. The target ref was there originally and it was moved down based on your comments - we need to have somewhere the split and "one of" choice.

If I am not getting something please let me know and show how your proposal would look like for other buffers (for example one pointing to a pod template)

jbtk replied:
Where top level? The min and max make only sense for buffer defined as a percent of the workload. Which top level do you mean and what would it mean there?

jackfrancis replied:
Is there any way we can drop the min/max entirely? I assume the purpose of this is to address the following:

- Referring to a deployment w/ 1 replica
- Maintain a +10% buffer

Ergo we don't actuate the buffer controller to accommodate addtional capacity for a new pod replica until the deployment replica count reaches 5 (assuming basic rounding up to the nearest int). Is this actually a problem in practice? In the case of a 10% Buffer that references a deployment pod spec that uses 1 core that hasn't yet replicated out to 5, we would still be guaranteeing node capacity for 100 millicores (assuming 1 replica).

tl;dr I'm not sure we really need the extra min/max. I understand that at small ints this might cause issues, but are we really solving for that use case? I think the canonical use case is the scale use case; or, if indeed there are deployment scaling use cases that we're solving for where the deployment replica flaps between small numbers, like 1 and 3, we can simply use a buffer with a high percent value like 100%.

Thoughts?

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+scheduling time will need to include the time of creating a new node and
+attaching it to the cluster. Many users care deeply about pod scheduling time
+and therefore would prefer to keep spare capacity in the cluster in order to
+speed up start time of the newly created pods.
```
Baseline capacity is not covered here yet as we removed the feature of having a buffer that is filled in by something for now. I will come back to it in another PR (updating this AEP) once this is approved so that we can start implementation of this initial, simple usecase.

x13n commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+    // user can specify per pod Requests to define the shape of the buffer space. In the latter case the
+    // CPU and Mem will be the maximum size of the buffer and if the Requests are too big to fit the total
+    // buffer size the admission of the buffer should be denied.
+    TotalCpu resource.Quantity
+    TotalMem resource.Quantity
```
GPUs are integers and assuming 1 GPU per chunk is not unreasonable. Having a `ResourceList` here would allow cloud provider specific chunking logic that would work for specific extended resources like GPU.

jbtk replied:
This would mean that our chunking logic would need to support whatever can be put in the ResourceList - at the moment this would be also storage. We could do this, but the question is - if we assume 1 GPU per chunk constant wouldn't it be better to define it in the per chunk resources? Also the more resources we have here the more weird chunking we can end up with (for example someone says that they want 1 GPU and 1000 CPU - this would not create a reasonable chunk - for memory and CPU it feels like there are less constraints like this, but maybe my intuition here is off.

x13n commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,574 @@
+
+We gathered feedback from the customers that maintenance of balloon pods and
+sizing them is something that should be simplified. Additionally if we have a
+buffer object we can add additional feature that are requested for example:
```
I think this is also doable with `CronJob` balloon pods, so not really an argument against them.

x13n commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,574 @@
+
+type NodeAutoscalingBufferStatus struct {
+    PodBufferCapacity   PodBufferCapacity
+	PodTemplateGeneration int	
```
nit: some indents on the API are not aligned well, I think because of mixing tabs with spaces.

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,576 @@
+        compared to running the balloons. For example: spare capacity size
+        depending on the deployment size or activating buffer depending on the
+        time of the day
+*   in the future cloud providers will be able to offer other types of buffers
```
Duplicate

x13n commented on cluster-autoscaler/proposals/buffers.md:
````diff
@@ -0,0 +1,576 @@
+
+```
+apiversion: autoscaling.x-k8s.io/v1alpha1
+kind: NodeAutoscalingBuffer
````
For me it would be 5, 7 or 8.

towca replied:
In order of preference: 7, 3, 8, 5, 2.

jackfrancis replied:
+1 for CapacityBuffer

jonathan-innis replied:
+1 for CapacityBuffer

towca commented:
> @jbtk thanks for addressing my comments! I feel like we're very close to alignment, let's finalize this so that we can start implementing something.
> 
> IIUC these are the contentious parts that are preventing us from approving the proposal:
> 
> 1. The `Buffer` name.
>    
>    * We need to distinguish that it's a Node-autoscaling-related (vs workload-autoscaling) object, so `Buffer` is too ambiguous.
>    * `NodeBuffer` suggests that the Buffer is "Node-centric" (i.e. "Buffer _consisting of_ Nodes"), while the semantics are actually more "workload-centric" (i.e. "Buffer _for a_ workload"). I agree that this is confusing and would prefer to avoid this option. `NodeHeadroom` has the same problem.
>    * Including `Reservation` in the name is IMO confusing because nothing will actually be "reserved" for Buffers until we implement the `fulfilledBy` semantics. Also there's the name collision with scheduler reservations that are being discussed separately, as @jbtk mentioned.
>    * So we need a name that is accurate, distinguishes from workload-autoscaling, doesn't suggest that it's "Node-centric", and doesn't have collisions with other k8s concepts. IMO `NodeAutoscalingBuffer`, or `ResourceBuffer` both satisfy all points (the accuracy part is of course subjective).
>    * IMO `CapacityBuffer` would be slightly better than `ResourceBuffer`. I'd also like to resurface my early suggestion of `node-autoscaling.x-k8s.io/Buffer` which also seems to satisfy all requirements.
> 2. Whether to include `TargetRef` in `ResourceBufferCapacity`.
>    
>    * @jbtk Do I understand correctly that `ResourceBufferCapacity` is supposed to mean "buffer for total amount of resources, and the arbitrary CRD just specifies where to create the resources"? If so, I'd name the field something else than `TargetRef` since the "target" here seems to be the total resources+chunk config (following the `ReplicaBufferCapacity` pattern where "target" is a workload). Maybe `NodeSelectorRef`/`NodeSelectorObject` or something similar?
>    * @ellistarn Having the reference to the object gives us better observability (we could use it in `kubectl` etc). It's also easier for the controller to identify the object to translate (imagine if e.g. we have different default chunking strategies depending on the object). NodeSelector can work for this, but it's imprecise and essentially hacking the semantics.
>    * Maybe it would be clearer if we put NodeSelector and the Ref in a separate subobject meant to select where to create the capacity? @ellistarn could you expand on what is your concern here?
> 3. The format of replica count/percentage in `ReplicaBufferCapacity`.
>    
>    * As mentioned in the other comment, for this one I'd defer to prior art in K8s API, or an opinion of a K8s API expert.
>    * Maybe we can get Tim to do another review pass?
> 4. Whether to include the `ProvisioningStrategy` field.
>    
>    * As @jbtk mentioned, we do have concrete use-cases in mind that we can talk about right now. The alternative is coming back to this discussion in a couple of months. I'd prefer to avoid that, as any iteration seems to have a non-trivial overhead of kicking off and maintaining the discussion.
>    * We can focus on the simple use-case that @jbtk mentioned - "stand-by"/"stopped" VMs that can be quickly resumed. Having the `ProvisioningStrategy` field allows the user to specify that the Buffer should be created using such VMs.
>    * @ellistarn I agree that we might need more discussion and design later when the use-cases are more concrete, but it seems safe to assume that we'll need to at least be able to distinguish from the default behavior. Which is all that the field does. Are you concerned that we'd want a completely separate API for this use-case?
> 
> A lot of this is subjective so it'd be good to hear more opinions. @gjtempleton @jackfrancis @x13n could you weigh in on the points listed above?

Based on the recent comments, there's another point of contention:

5. Do we need the controller part for translating a Buffer Spec into the virtual Pods that autoscalers inject?
    * The alternative is to have the translation as part of Node autoscaler responsibilities, leveraging a shared library for the implementation.
    * A hard requirement for the translation is extensibility. Regardless of the solution, we need to be able to translate cloud-provider-specific objects on top of the shared part. We can do that both via a controller and a library with similar ease.
    * Using the controller approach we could move the translation logic out of Node autoscalers to reduce their overall scope. We could also have users run controllers for translating their own custom objects. I agree that both these points are pretty speculative, the first iteration wouldn't be utilizing either.
    * When we take quotas into account, the controller approach allows us to similarly not have the quota logic in Node autoscalers. This is out of scope for this first iteration, but it should be a fast follow and probably a requirement for beta - so I wouldn't call that part speculative. The quota logic is also likely to be more involved and have less affinity to Node autoscalers than the translation logic.
    * We've decided on referencing PodTemplate objects instead of inlining them on the advice of k8s API experts. So if we don't go with the controller approach but still want to have the pod spec in the status for observability, the shared library would have to create PodTemplate objects. This is pretty natural for a controller, but seems weird for such a library.
    * IMO it's pretty subjective which solution is "cleaner" or incurs less cognitive burden. I see the argument for "the proposal is less complex when we take the controller part out" (this was my first intuition as well), but on the other hand I definitely see the argument for "the simpler proposal without controller further bloats Node autoscaler scope, resulting in a more complex end-state". The ever-expanding scope is one of the biggest problems we see in CA, so delegating new not-strictly-autoscaling responsibilities to other components (even if it's just separating them under the same binary to start with) seems worthwhile.
    * IMO out of all the above, the quota part seems the most convincing. We'll have to implement quota logic soon, and I'd strongly prefer not to have to add it to CA directly.

jbtk commented:
For the controller discussion there is one more argument. For buffers for objects implementing scale subresource if we go with a library and no pod template in the status we would be able to buffer capacity only if at least one pod is running in the cluster. This is a limitation for people who want buffers for jobs that run only some time during a day.

@ellistarn @towca @jonathan-innis  as FYI

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,571 @@
+In case of the cluster autoscaler we would deploy the controller as an optional
+subprocess:
+
+![Figure](./images/buffers-component-structure.jpg)
```
The original intention of the proposal was a controller running within the ca process.

So that anyone can turn on this processing in CA and turn off if they would rather provide their own controller. The idea is that it would be also possible for them to deploy their own controller if they want to handle buffers that are not handled by the in CA controller (reusing likely some parts because of quota integration etc.). 

The drawing has the component rectangle on top, but change the wording to make it clear that the buffer controller is a loop within the autoscaler*

* and is built using a library that is easy to reuse

jonathan-innis commented:
@jbtk and I had a conversation this morning -- there's been some more thought put into the API design and use-cases and I'm summarizing some of the decisions that we took:
1. We settled on `CapacityBuffer` as the kind naming — we’ll go forward with this in the AEP, it disambiguates what Buffer means
2. We shifted from having the pod template inlined in either the spec or the status to using the PodTemplate object present in K8s. There’s a couple of reasons for this: First, this comes pre-packaged with validations on the apiserver rather than having to copy over every part of the spec over into our API (this was also recommended from since there’s a push against inlining PodSpec for newer APIs). Second, this allows us to show what the resolved spec is for the buffer without muddying the spec too too much — so you can get visibility if you want into the pod that we are using for modeling, but don’t actually need it in the BufferCapacity Spec
3. We agreed that `ProvisioningStrategy` was reasonable to keep since it could have values like Default or WarmPool as ways to handle the capacity in the future when users want to opt for non-active capacity to save money
4. I proposed two versions of the API that I’m personally comfortable with when it comes to naming, symmetry, and avoiding over-nesting (I’m personally good with either of these options to be the proposed design)

API 1 (separate stanzas for validation for podTemplate and scaleTarget)
```yaml
kind: CapacityBuffer
spec:
  podTemplate:
    objectRef:
    replicas:
  scaleTarget: # HPA uses same naming 
    objectRef:
      
    percentage:
    replicas:
  provisioningStrategy: <>
  limits:
status:
  conditions:
  podTemplateRef:
    name: 
  podTemplateGeneration:
  replicas:
```

API 2 (flattened API)
```yaml
kind: CapacityBuffer
spec:
  podTemplateRef:
    name:
  scaleTargetRef:
    name:
  percentage:
  replicas:
  limits:
   
status:
  conditions:
  podTemplateRef:
    name: 
  podTemplateGeneration:
  replicas:
```

jbtk commented:
Updated the proposal to reflect the changes. The field names slightly differ from what @jonathan-innis proposed - please review and comment directly on the pr.

Note: While discussing with @jonathan-innis he noticed that there might be security considerations to reference across namespaces. For now I changed the pod template to be string (name only) within the same namespsce and I am discussing options with @liggitt. Open option - adding a reference grant to allow cross namespace referencing.

ellistarn commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,525 @@
+type CapacityBufferSpec struct {
+  // optional
+  // oneof=Capacity
+  ScaleSubresourceCapacity *ScaleSubresourceBufferCapacity
```
regarding naming, one of the patterns I try to avoid is "stuttering". Each parent name should set the context, allowing for the naming to flow when reading. e.g.

```
scaleSubresourceCapacity:
  scaleTargetRef:
```

We stutter both `capacity` (the name of the API) and `scale` with this approach. 

You could simply with 

```
scaleSubresourceCapacity:
  targetRef
```
or even
```
scaleSubresourceCapacity:
  ref
```

Separately, I'm finding that `scaleSubresourceCapacity` is quite verbose in-and-of-itself, partially due to the `capacity`. It's likely that users are less familiar with what a scale subresource is, as well, even if it's technically accurate. Something like:
```
target:
  ref:
```
communicates more or less the same information with less cognitive overhead.

---

However, I recognize that we're aiming for a "one-of" semantic for each variant. Does this naming make sense in that context?

PodCapacity shares a similar stutter of `capacity` with capacity buffer. 

Something like:

```
podTemplate:
  replicas: 5
  name: foo
---
target:
  percentage: 
  replicas: 5
  ref:
    kind: Deployment
    name: bar
```
has nice parity. In this case, the two options of capacity buffers are "target capacity buffers" and "pod template capacity buffers". 

---

Finally, it feels a bit awkward that `limits` is a top level field, but `replicas` and `percentage` are confined to their individual stanzas. Replicas also duplicated in both stanzas. The only special case is `percentage`, which only makes sense with scalable objects. This is what leans me towards @jonathan-innis's other proposal in slack with a more collapsed structure.

```
limits: ...
replicas: ... # oneof with percentage
percentage: ... # oneof with percentage
targetRef:  # ScaleSubresource OR PodTemplate
  kind: ...
  name: ...
```

Yes, this does mean slightly more validation rules (i.e., you cannot set percentage with a PodTemplate target).

---

That's enough bike shedding. But those are my thoughts.

jbtk replied:
@liggitt @thockin - you happened to look before at this API - it changed a bit, but it seems that we finally aligned on the features and what the API does, but we have still some fine tuning to do on the naming and structure. Are there any best practices regarding:
- naming of the fields and types - should they repeat the context or assume that parent gives the context already (see comment above from @ellistarn about stuttering)
- do we prefer to have more structured yaml with more nesting, but also grouping of fields or do we prefer more flattened structure (resulting in simpler yaml structure) and rely on comments and documentation for the users to understand what combination of fields should/can be used
- also any other field naming ideas very welcome as due to generic nature of words like "buffer" and "capacity" naming in this api was a challenge for me

As for the scaleSubresourceCapacity - this is one of the fields that is for me the worst name in the whole API. I do not know how to name it well and at the same time have simple and easy to understand name - I have a really hard time explaining what is it other than something that implements scale subresource. With target we are basically assuming that the user needs to read documentation to even know what to put there. If we were to flatten I would rather go with the last proposal from @ellistarn where the target can be either scale subresource or pod template. My main concern here would be though that if a new workload definition is introduced (as worked on in: ) the will be yet another target type and the params like replicas and percentage may not be applicable any more and then we need to add all of them to top level to have the api consistent. Therefore I was trying to structure the API to represent the logical bits. But I know that in the end we read the configs more than write so maybe simpler yaml is worth it? 

Maybe others want to chime in - @towca @jackfrancis. I am happy to follow best practices or any existing patterns.

towca replied:
My opinions on the points above:
* The stutter seems unavoidable in any name that is a top-level type. In order not to pollute the API "namespace", the names of the types should all be clearly tied to the main `CapacityBuffer` name. Given this, I'm not sure if we can find better names for top-level types than the ones currently proposed.
* Given the required verbosity of the type names, I would avoid repeating the stutter in the field names wherever possible. The proposed field names already remove the `Buffer` part, removing the `Capacity` part too would leave us with just `ScaleSubresource` and `Pod` as field names - which is IMO too vague even in context. Out of all the field names, `ScaleTargetRef` seems like the only one that can be shortened to `TargetRef` or `Ref` without decreasing readability.
* Nesting the different "capacity specs" seems better to me than flattening the structure completely. Even though some of the parameters like `Replicas` are repeated, IMO it's a very worthwhile tradeoff for the clarity of which parameters are applicable to which spec. And as Justyna mentions, we might have capacity types in the future for which even the `Replicas` field wouldn't be applicable. So on one hand we have an API that is slightly easier to read and write once you understand it, and on the other one that is easier to understand and extend - IMO the latter wins.

In any case, the current proposal LGTM. @jackfrancis I'm curious to hear your thoughts on these trade-offs, they are the last bits that we need to align on.

jbtk replied:
Applied the suggestion to the PR.

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,525 @@
+  // +optional
+  Replicas *int
+  // +optional
+  Percentage *int
```
Done.

jbtk commented:
@ellistarn @towca 

I changed the names and yaml structure based on the comment from @liggitt (thanks!). 

Are there any more things that are blocking this PR from being approved for merge?

ellistarn commented:
> @ellistarn @towca 
> 
> I changed the names and yaml structure based on the comment from @liggitt (thanks!). 
> 
> Are there any more things that are blocking this PR from being approved for merge?

Lgtm

towca commented:
The current proposal LGTM. We got an LGTM from Karpenter and Jordan as well. IMO this is good to merge as soon as the remaining nits from me and Jordan are addressed.

/lgtm

jackfrancis commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,540 @@
+
+### Story 1
+
+The user has a deployment of pods that are serving user traffic. In order to
```
It might be worth mentioning in this user story the importance of optimizing the way we do this. In the provided example of 10 percent, we are changing the current behavior from this:

- create n nodes in reponse to the sum of required infra that can schedule Pending pods

to this:

- create n nodes in reponse to the sum of required infra that can schedule Pending pods +
- create n nodes in reponse to (the sum of required infra that can schedule Pending pods) * .10

We want to ensure that our plan includes for a way to accommodate the new buffer scenario with a single scale-out event, rather than 

- scale out by 10 new nodes
- scale out by 1 new node

(or worse, a serial scale out where we get 1 node, and *then* 10 nodes)

jbtk replied:
We will use the status of scale subresource so we will scale to existing replicas. So always they will show up first before we adjust the size of the buffer.
We may run into scale out by 10 and then 1 (as there is a buffer controller loop that needs to process the change), but since we will always prioritize shards with actual pods the scale up for buffer is ok to happen.

Added a note so that this is captured and not forgotten about as I did not though about this edgecase.

jackfrancis commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,540 @@
+type CapacityBufferSpec struct {
+  // optional
+  // oneof=PodShapeSource
+  PodTemplateRef *LocalObjectRef
```
for `ScalableRef ` we have the comment:

`// Reference to an object of a kind that has scale subresource and sets its label selector field.`

Let's add a similar comment for `PodTemplateRef`, maybe something like:

`// Reference to a pre-defined PodTemplate resource in the same namespace that declares a static buffer`

jackfrancis commented:
I added some additional nits, but overall I'm happy with the direction of the API that represents a consensus between @ellistarn @liggitt @towca (and others). lgtm after incorporation of that last mile feedback

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,540 @@
+  namespace: my-namespace
+spec:
+  scalableRef
+    kind: Deployment
```
Thanks, updated.

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,540 @@
+
+  // If empty it will create additional nodes to provide capacity in the cluster.  
+  // Cloud providers can offer their own buffering strategies.
+  ProvisioningStrategy *string
```
Yep, this is right - updated.

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,540 @@
+  Limits *ResourceList
+}
+
+type PodBufferCapacity struct {
```
Yes, removed.

jbtk commented on cluster-autoscaler/proposals/buffers.md:
```diff
@@ -0,0 +1,540 @@
+in autoscaling space will make it clear and will allow for releasing it
+independently of the core k8s.
+
+## Features to be added in next iterations
```
Done.

jbtk commented:
/unhold
